### PR TITLE
paych actor tests

### LINKED ISSUE #452: Create Payment Channel Actor Tests
**Task summary**

Write tests for the Payment Channel Actor. 

Similar to the other tests that have been done they can be found under forest/vm/actor/tests

Relevant Actor : forest/vm/actor/src/builtin/paych

**Specification reference**

-  Lotus Payment Channel Actor Test : 
https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/paych/paych_test.go

### PR DESCRIPTION
**Pull request by nannick**
**Summary of changes**
Changes introduced in this pull request:
-  Added tests to paych Actor
-  Minor updates in paych actor logic
-  Introduced some Syscall Trait implementation for MockRuntime

**Reference issue to close (if applicable)**

Closes #452 

**Other information and links**

### CODE REVIEW

File: vm/actor/tests/paych_actor_test.rs [hunk 1]
```diff
@@ -0,0 +1,1170 @@
...
+
+    struct TestCase {
+        paych_addr: Address,
+        caller_code: Cid,
+        new_actor_code: Cid,
+        payer_code: Cid,
+        expected_exit_code: ExitCode,
+    }
+
+    let test_cases: Vec<TestCase> = vec![TestCase {
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/actor/tests/paych_actor_test.rs:103]: Oh it looks like you havent implemented some of the other test cases here, for example if the paych_addr is not an ID type

File: vm/actor/src/builtin/paych/mod.rs [hunk 2]
```diff
@@ -349,6 +348,10 @@ impl ActorCode for Actor {
                 Self::collect(rt)?;
                 Ok(Serialized::default())
             }
+            Some(Method::UpdateChannelState) => {
+                Self::update_channel_state(rt, params.deserialize().unwrap())?;
```
Comment (dutterbutter) [inline vm/actor/src/builtin/paych/mod.rs:352]: replace `unwrap()` with `?`

File: vm/actor/tests/common/mod.rs [hunk 3]
```diff
@@ -9,15 +9,20 @@ use actor::{
...
+use crypto::{DomainSeparationTag, Signature};
+use encoding::blake2b_256;
 use encoding::{de::DeserializeOwned, Cbor};
+use fil_types::{PieceInfo, RegisteredProof, SealVerifyInfo, WindowPoStVerifyInfo};
 use ipld_blockstore::BlockStore;
 use message::{Message, UnsignedMessage};
 use runtime::{ActorCode, Runtime, Syscalls};
 use std::cell::{Cell, RefCell};
 use std::collections::{HashMap, VecDeque};
 use vm::{ActorError, ExitCode, MethodNum, Randomness, Serialized, TokenAmount};
```
Comment (dutterbutter) [inline vm/actor/tests/common/mod.rs:22]: Nit: remove whitespace

File: vm/actor/tests/common/mod.rs [hunk 4]
```diff
@@ -125,6 +145,26 @@ impl<'a, BS: BlockStore> MockRuntime<'a, BS> {
         assert!(addr.len() > 0, "addrs must be non-empty");
         *self.expect_validate_caller_addr.borrow_mut() = Some(addr.to_vec());
     }
+    pub fn expect_validate_caller_type(&self, ids: &[Cid]) {
+        assert!(ids.len() > 0, "addrs must be non-empty");
+        *self.expect_validate_caller_type.borrow_mut() = Some(ids.to_vec());
+    }
+
+    #[allow(dead_code)]
```
Comment (dutterbutter) [inline vm/actor/tests/common/mod.rs:153]: I think you can remove this

File: vm/actor/tests/common/mod.rs [hunk 5]
```diff
@@ -509,11 +558,77 @@ impl<BS: BlockStore> Runtime<BS> for MockRuntime<'_, BS> {
         todo!("implement me???")
     }
 
+    fn total_fil_circ_supply(&self) -> Result<TokenAmount, ActorError> {
```
Comment (dutterbutter) [inline vm/actor/tests/common/mod.rs:561]: You will have some conflicts when my miner actor impl comes in. Just fyi.

File: vm/actor/tests/reward_actor_test.rs [hunk 6]
```diff
@@ -12,15 +12,17 @@ use common::*;
 use db::MemoryDB;
 use ipld_blockstore::BlockStore;
 use message::UnsignedMessage;
+
```
Comment (dutterbutter) [inline vm/actor/tests/reward_actor_test.rs:15]: Nit: remove white space

File: vm/actor/src/builtin/paych/mod.rs [hunk 7]
```diff
@@ -93,12 +93,11 @@ impl Actor {
...
         };
 
-        let mut sv = params.sv;
-
+        let sv = params.sv;
         // Pull signature from signed voucher
         let sig = sv
             .signature
-            .take()
+            .clone()
```
[the remark this answers is not in the document]
Comment (nannick) [reply vm/actor/src/builtin/paych/mod.rs:100]: The issue before is the signature got moved out which affected how the sv got represented in bytes.  The as_ref is a better fix

File: vm/actor/tests/paych_actor_test.rs [hunk 8]
```diff
@@ -0,0 +1,1170 @@
...
+                .build()
+                .unwrap(),
+            TestCase::builder()
+                .sig(sig.clone())
+                .secret_preimage(vec![0; 2 << 21])
+                .build()
+                .unwrap(),
+        ];
+
+        for test_case in test_cases {
```
[the remark this answers is not in the document]
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:228]: Yeah in the Lotus they had a description for each test case. I'll print it out before each test executes.

File: vm/actor/tests/paych_actor_test.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:103]: Whoops my bad missed that case

File: vm/actor/tests/paych_actor_test.rs [hunk 9]
```diff
@@ -0,0 +1,1192 @@
...
+    let merge_from: &LaneState = &state.lane_states[1];
+    sv.lane = merge_to.id;
+    let merge_nonce = merge_to.nonce + 10;
+    let merges: Vec<Merge> = vec![Merge {
+        lane: merge_from.id,
+        nonce: merge_nonce,
+    }];
+
+    sv.merges = merges;
+    let payee_addr = Address::new_id(103);
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:467]: Can you create constants for each of these, so you're not just referencing a random address for each. (For example this one is the payee address)

File: vm/actor/tests/paych_actor_test.rs [hunk 10]
```diff
@@ -0,0 +1,1192 @@
...
+    let lane_vec = vec![1, 1, 1, 0];
+    let voucher_vec = vec![10, 0, 10, 10];
+    let balance_vec = vec![0, 0, 1, 0];
+    let merge_vec = vec![1, 10, 10, 10];
+    let exit_vec = vec![
+        ExitCode::ErrIllegalArgument,
+        ExitCode::ErrIllegalArgument,
+        ExitCode::ErrIllegalState,
+        ExitCode::ErrIllegalArgument,
+    ];
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:528]: Why not just create a struct and have a 4 len array of them so you don't have to index each like this

File: vm/actor/tests/paych_actor_test.rs [hunk 11]
```diff
@@ -0,0 +1,1192 @@
...
+
+    let exp_settling_at = SETTLE_DELAY + 10;
+    let state: PState = rt.get_state().unwrap();
+    assert_eq!(exp_settling_at, state.settling_at);
+    assert_eq!(state.min_settle_height, 0);
+
+    let min_settle_vec = vec![0, 2, 12];
+    let exp_min_settle_height = vec![state.min_settle_height, 2, 12];
+    let exp_settling_at = vec![state.settling_at, state.settling_at, 12];
+    let num_test_cases = min_settle_vec.len();
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:769]: again here, the indexing is a bit unweildy to use in future

File: vm/actor/tests/paych_actor_test.rs [hunk 12]
```diff
@@ -0,0 +1,1192 @@
...
+        rt.expect_validate_caller_addr(&[state.from, state.to]);
+        rt.expect_verify_signature(
+            ucp.sv.clone().signature.unwrap(),
+            state.to,
+            to_vec(&ucp.sv).unwrap(),
+            ExitCode::Ok,
+        );
+        assert!(rt
+            .call(
+                &PAYCH_ACTOR_CODE_ID.clone(),
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:787]:
```suggestion
                &*PAYCH_ACTOR_CODE_ID,
```
I didn't verify in code, but you should be able to do this for these

File: vm/actor/tests/paych_actor_test.rs [hunk 13]
```diff
@@ -0,0 +1,1192 @@
...
+#[test]
+fn succeed_correct_secret() {
+    let bs = MemoryDB::default();
+    let (mut rt, sv) = require_create_cannel_with_lanes(&bs, 1);
+    let state: PState = rt.get_state().unwrap();
+    let ucp = UpdateChannelStateParams {
+        proof: vec![],
+        secret: vec![],
+        sv: sv.clone(),
+    };
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:807]: just an idea (probably not worth doing), but since this is constructed so many times, idea could be to have a `From<SignedVoucher>` impl so you can construct from sv as is done many times

File: vm/actor/tests/paych_actor_test.rs [hunk 14]
```diff
@@ -0,0 +1,1192 @@
...
+#[test]
+fn actor_collect() {
+    let dont_settle_vec = vec![true, false, false];
+    let exp_send_from_vec = vec![ExitCode::Ok, ExitCode::ErrPlaceholder, ExitCode::Ok];
+    let exp_send_to_vec = vec![ExitCode::Ok, ExitCode::Ok, ExitCode::ErrPlaceholder];
+    let exp_collect_vec = vec![
+        ExitCode::ErrForbidden,
+        ExitCode::ErrPlaceholder,
+        ExitCode::ErrPlaceholder,
+    ];
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:981]: again here, very hard to follow these

File: vm/actor/tests/paych_actor_test.rs [hunk 15]
```diff
@@ -0,0 +1,1192 @@
...
+        time_lock_max: u64::MAX,
+        secret_pre_image: vec![],
+        extra: Option::None,
+        lane: param.lane,
+        nonce: param.nonce,
+        amount: BigInt::from(param.amt),
+        min_settle_height: 0,
+        merges: vec![],
+        signature: Some(sig.clone()),
+    };
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:1104]: SignedVoucher can probably have a default implementation derived, to avoid having to construct all of this when you are just setting a few values, and just defaulting all ignored fields with `..Default::default()`

File: vm/actor/tests/paych_actor_test.rs [hunk 16]
```diff
@@ -0,0 +1,1192 @@
...
+    receiver: Address,
+) {
+    let state: PState = rt.get_state().unwrap();
+    let expected_state = PState::new(sender, receiver);
+    assert_eq!(expected_state.to, state.to);
+    assert_eq!(expected_state.from, state.from);
+    assert_eq!(expected_state.min_settle_height, state.min_settle_height);
+    assert_eq!(expected_state.settling_at, state.settling_at);
+    assert_eq!(expected_state.to_send, state.to_send);
+    assert_eq!(state.lane_states.len(), 0);
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:1171]: You're not calling verify_state here, and it seems like you should? (Seems like some of your tests may be wrong, as  `TestActor_UpdateChannelStateRedeem/redeeming_voucher_updates_correctly_with_one_lane` and `TestActor_UpdateChannelStateMergeSuccess` equivalents should have failed with this setup)

Seems merge success test doesn't even call verify_state and it looks like it should?

File: vm/actor/tests/paych_actor_test.rs (hunk 10 above)
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:528]: Yeah idk what I was thinking tbh it wouldve been less work to put in a actual struct

File: vm/actor/tests/paych_actor_test.rs (hunk 16 above)
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:1171]: Yeah I just manually inlined the function since it was a fixed call with -1. Shouldve realised it made it harder to compare with Lotus

File: vm/actor/tests/paych_actor_test.rs (hunk 12 above)
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:787]: Yup changed added

File: vm/actor/tests/paych_actor_test.rs (hunk 13 above)
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:807]: I never implemented this trait before might as well lol

File: vm/actor/tests/paych_actor_test.rs (hunk 9 above)
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:467]: I can do it for the majority of the values but they arent consistent with the variable names and values.

In one they set paych_addr to Address::new_id(100) and in another its  Address::new_id(101)

File: vm/actor/tests/paych_actor_test.rs (hunk 15 above)
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:1148]: ye should shave off a few lines

File: vm/actor/tests/paych_actor_test.rs (hunk 9 above)
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs]: Best solution is to probably put them in diff mods should be in my next commit

File: vm/actor/tests/paych_actor_test.rs [hunk 17]
```diff
@@ -0,0 +1,1265 @@
...
+                .unwrap_err();
+            assert_eq!(tc.exp_send_collect, error.exit_code());
+        }
+    }
+}
+
+struct ActorHarness {
+    payer: Address,
+    payee: Address,
+}
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:1118]: This isn't needed, there was no need to add this complexity to it as these are always just static values

File: vm/actor/tests/common/mod.rs [hunk 18]
```diff
@@ -125,6 +142,26 @@ impl<'a, BS: BlockStore> MockRuntime<'a, BS> {
         assert!(addr.len() > 0, "addrs must be non-empty");
         *self.expect_validate_caller_addr.borrow_mut() = Some(addr.to_vec());
     }
+    pub fn expect_validate_caller_type(&self, ids: &[Cid]) {
```
Comment (austinabell) [inline vm/actor/tests/common/mod.rs:145]: can you add a dead code tag on this? not used for other tests and throwing warnings

File: vm/actor/tests/common/mod.rs [hunk 19]
```diff
@@ -512,11 +565,73 @@ impl<BS: BlockStore> Runtime<BS> for MockRuntime<'_, BS> {
...
+    fn verify_signature(
+        &self,
+        signature: &Signature,
+        signer: &Address,
+        plaintext: &[u8],
+    ) -> Result<(), Box<dyn StdError>> {
+        let op_exp = self.expect_verify_sig.replace(Option::None);
+
+        if let Some(exp) = op_exp {
+            if exp.sig == *signature && exp.signer == *signer && (exp.plaintext[..]) == *plaintext {
```
Comment (austinabell) [inline vm/actor/tests/common/mod.rs:590]:
```suggestion
            if exp.sig == *signature && exp.signer == *signer && &exp.plaintext[..] == plaintext {
```
Just a bit weird to read as is, could be more explicit with `.as_slice()` too for comparison

File: vm/actor/tests/paych_actor_test.rs [hunk 20]
```diff
@@ -0,0 +1,1115 @@
...
+) {
+    assert_eq!(
+        exp,
+        rt.call(&*PAYCH_ACTOR_CODE_ID, method_num, ser,)
+            .unwrap_err()
+            .exit_code()
+    );
+}
+
+mod paych_constructor {
```
Comment (austinabell) [inline vm/actor/tests/paych_actor_test.rs:55]: why do you put tests in seperate modules? doesn't do anything afaik?

Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:55]: Well its only really to mimic the format they have in GO where they can run multiple tests in a func. Other than that by keeping them in diff modules helps put funcs/constants that are used in certain tests together

File: vm/actor/tests/paych_actor_test.rs [hunk 21]
```diff
@@ -0,0 +1,1192 @@
...
+        assert_eq!(test_case.expected_exit_code, error.exit_code());
+    }
+}
+
+mod create_lane_tests {
+    use super::*;
+    #[derive(Builder, Debug)]
+    #[builder(name = "TestCaseBuilder")]
+    struct TestCase {
+        desc: String,
```
[the remark this answers is not in the document]
Comment (nannick) [reply vm/actor/tests/paych_actor_test.rs:174]: Ill leave it as is then

File: vm/actor/tests/common/mod.rs [hunk 22]
```diff
@@ -126,6 +143,12 @@ impl<'a, BS: BlockStore> MockRuntime<'a, BS> {
         *self.expect_validate_caller_addr.borrow_mut() = Some(addr.to_vec());
     }
 
+    #[allow(dead_code)]
+    pub fn expect_verify_signature(&self, exp: ExpectedVerifySig) {
+        //assert!(addr.len() > 0, "addrs must be non-empty");
```
Comment (dutterbutter) [inline vm/actor/tests/common/mod.rs:148]: Remove commented code.
